lib/logmsg/nvtable-serialize.c

```diff
@@ -24,199 +24,269 @@
+{
+  if (_is_static_entry(entry))
+    return old_handle;
+  else
+    return log_msg_get_value_handle(nv_entry_get_name(entry));
 }
 
-void
-_copy_handles(NVHandle* handles_to_update, NVHandle* new_updated_handles, guint8 num_handles_to_update)
+static NVHandle
+_allocate_handle_of_referenced_entry(NVTable *self, NVHandle ref_handle)
 {
-  guint16 i;
-  for (i = 0; i < num_handles_to_update; i++)
-    {
-      handles_to_update[i] = new_updated_handles[i];
-    }
+  NVIndexEntry *index_entry;
+  NVEntry *ref_entry = nv_table_get_entry(self, ref_handle, &index_entry);
+
+  return _allocate_handle_for_entry_name(ref_handle, ref_entry);
 }
 
-const gchar *
-_get_entry_name(NVTable *self, NVDynValue *dyn_entry)
+static gboolean
+_is_indirect(NVEntry *entry)
 {
-  NVEntry *entry = nv_table_get_entry_at_ofs(self, dyn_entry->ofs);
-  if (!entry)
-    return NULL;
-  return nv_entry_get_name(entry);
+  return entry && entry->indirect;
 }
 
 static void
-_set_updated_handle(NVDynValue *dyn_entry, NVHandle new_handle, NVHandle *handles_to_update, NVHandle *updated_handles, guint8 num_handles_to_update)
+_fixup_handle_in_indirect_entry(NVTable *self, NVEntry *entry)
 {
-  NVHandle old_handle = dyn_entry->handle;
-  guint8 j;
-  dyn_entry->handle = new_handle;
-  if (handles_to_update)
-    {
-      for (j = 0; j < num_handles_to_update; j++)
-        {
-          if (handles_to_update[j] == old_handle)
-            {
-              updated_handles[j] = new_handle;
-            }
-        }
-    }
+  entry->vindirect.handle = _allocate_handle_of_referenced_entry(self, entry->vindirect.handle);
 }
 
-static void
-_update_dynamic_handles(NVTable *self, NVRegistry *logmsg_nv_registry,
-                        NVHandle *handles_to_update,
-                        guint8 num_handles_to_update)
+static gboolean
+_validate_entry(LogMessageSerializationState *state, NVEntry *entry)
 {
-  guint16 i;
-  NVHandle *updated_handles = g_new0(NVHandle, num_handles_to_update);
-  NVDynValue* dyn_entries = nv_table_get_dyn_entries(self);
+  NVTable *nvtable = state->nvtable;
 
-  for (i = 0; i < self->num_dyn_entries; i++)
-    {
-      NVHandle new_handle;
-      NVDynValue *dyn_entry = &dyn_entries[i];
-      const gchar *name = _get_entry_name(self, dyn_entry);
+  /* check alignment */
+  if ((GPOINTER_TO_UINT(entry) & 0x3) != 0)
+    return FALSE;
 
-      if (!name)
-        continue;
+  /* entry points above the start of the NVTable */
+  if (GPOINTER_TO_UINT(entry) < GPOINTER_TO_UINT(nvtable))
+    return FALSE;
 
-      new_handle = nv_registry_alloc_handle(logmsg_nv_registry, name);
+  /* entry header is inside the allocated NVTable */
+  if (GPOINTER_TO_UINT(entry) + NV_ENTRY_DIRECT_HDR > GPOINTER_TO_UINT(nvtable) + nvtable->size)
+    return FALSE;
+  /* entry as a whole is inside the allocated NVTable */
+  if (GPOINTER_TO_UINT(entry) + entry->alloc_len > GPOINTER_TO_UINT(nvtable) + nvtable->size)
+    return FALSE;
```

Comment:
I think these checks, validations should be in separated static inline functions, with good names and in that case we have not need for the comments.
But this is a very minor note